examples/remarketing/upload_conversion_enhancement.rb

```diff
@@ -0,0 +1,202 @@
+#
+# Adjusts an existing conversion by supplying user identifiers so Google can
+# enhance the conversion value.
+
+require 'optparse'
+require 'google/ads/google_ads'
+require 'digest'
+
+# [START upload_conversion_enhancement]
+def upload_conversion_enhancement(
+  customer_id,
+  conversion_action_id,
+  order_id,
+  conversion_date_time,
+  user_agent,
+  restatement_value,
+  currency_code
+)
+  # GoogleAdsClient will read a config file from
+  # ENV['HOME']/google_ads_config.rb when called without parameters
+  client = Google::Ads::GoogleAds::GoogleAdsClient.new
+
+  # [START create_adjustment]
+  enhancement = client.resource.conversion_adjustment do |ca|
+    ca.conversion_action = client.path.conversion_action(customer_id, conversion_action_id)
+    ca.adjustment_type = :ENHANCEMENT
+    ca.order_id = order_id
+
+    unless conversion_date_time.nil?
+      ca.gclid_date_time_pair = client.resource.gclid_date_time_pair do |pair|
+        pair.conversion_date_time = conversion_date_time
+      end
+    end
+
+    ca.user_identifiers << client.resource.user_identifier do |ui|
+      ui.address_info = client.resource.offline_user_address_info do |info|
+        info.hashed_first_name = normalize_and_hash("Joanna")
+        info.hashed_last_name = normalize_and_hash("Smith")
+        info.hashed_street_address = normalize_and_hash("1600 Amphitheatre Pkwy")
+        info.city = "Mountain View"
+        info.state = "CA"
+        info.postal_code = "94043"
+        info.country_code = "US"
+      end
+      ui.user_identifier_source = :FIRST_PARTY
+    end
+
+    ca.user_identifiers << client.resource.user_identifier do |ui|
+      ui.hashed_email = normalize_and_hash_email("joannasmith@example.com")
+      ui.user_identifier_source = :FIRST_PARTY
+    end
+
+    unless user_agent.nil?
+      # Sets the user agent. This should match the user agent of the request
+      # that sent the original conversion so the conversion and its enhancement
+      # are either both attributed as same-device or both attributed as
+      # cross-device.
+      ca.user_agent = user_agent
+    end
+
+    unless restatement_value.nil?
+      ca.restatement_value = client.resource.restatement_value do |ra|
+        ra.adjusted_value = restatement_value.to_f
+        unless currency_code.nil?
+          ra.currency_code = currency_code
+        end
+      end
+    end
+  end
+  # [END create_adjustment]
+
+  response = client.service.conversion_adjustment_upload.upload_conversion_adjustments(
+    customer_id: customer_id,
+    conversion_adjustments: [enhancement],
+    partial_failure: true,
+  )
+
+  if response.partial_failure_error
+    puts "Partial failure encountered: #{response.partial_failure_error.message}."
+  else
+    result = response.results.first
+    puts "Uploaded conversion adjustment of #{result.conversion_action} for "\
+      "order ID #{result.order_id}."
+  end
+end
+# [END upload_conversion_enhancement]
+
+
+def normalize_and_hash(str)
```

Comment:
Please add START/END for `normalize_and_hash` from:

https://github.com/googleads/google-ads-java/blob/main/google-ads-examples/src/main/java/com/google/ads/googleads/examples/remarketing/UploadConversionEnhancement.java#L269